File: server/src/main/java/org/dreamexposure/discal/server/utils/Authentication.java
package org.dreamexposure.discal.server.utils;

import com.google.api.client.http.HttpStatusCodes;

import org.dreamexposure.discal.core.database.DatabaseManager;
import org.dreamexposure.discal.core.logger.LogFeed;
import org.dreamexposure.discal.core.logger.object.LogObject;
import org.dreamexposure.discal.core.object.BotSettings;
import org.dreamexposure.discal.core.object.web.AuthenticationState;
import org.dreamexposure.discal.core.object.web.UserAPIAccount;
import org.dreamexposure.discal.core.utils.GlobalConst;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;

import javax.servlet.http.HttpServletRequest;

public class Authentication {
    private static Timer timer;

    private static final Map<String, Long> tempKeys = new HashMap<>();
    private static final Map<String, Long> readOnlyKeys = new HashMap<>();

    public static AuthenticationState authenticate(final HttpServletRequest request) {
        if (!"POST".equalsIgnoreCase(request.getMethod()) || "GET".equalsIgnoreCase(request.getMethod())) {
            LogFeed.log(LogObject.forDebug("Denied Access", "Method: " + request.getMethod()));
            return new AuthenticationState(false)
                .status(GlobalConst.STATUS_NOT_ALLOWED)
                .reason("Method not allowed");
        }
        //Requires "Authorization Header
        if (request.getHeader("Authorization") != null) {
            final String key = request.getHeader("Authorization");

            //Check if this is from within the DisCal network...
            if (key.equals(BotSettings.BOT_API_TOKEN.get())) {
                return new AuthenticationState(true)
                    .status(GlobalConst.STATUS_SUCCESS)
                    .reason("Success")
                    .keyUsed(key)
                    .fromDisCalNetwork(true);
                //Check if this is a temp key, granted for a logged in user...
            } else if (tempKeys.containsKey(key)) {
                return new AuthenticationState(true)
                    .status(GlobalConst.STATUS_SUCCESS)
                    .reason("Success")
                    .keyUsed(key)
                    .fromDisCalNetwork(false)
                    .readOnly(false);
            } else if (readOnlyKeys.containsKey(key)) {
                return new AuthenticationState(true)
                    .status(GlobalConst.STATUS_SUCCESS)
                    .reason("Success")
                    .keyUsed(key)
                    .fromDisCalNetwork(false)
                    .readOnly(true);
            } else if ("teapot".equals(key)) {
                return new AuthenticationState(false)
                    .status(GlobalConst.STATUS_TEAPOT)
                    .reason("I'm a teapot")
                    .keyUsed(key);
            }

            //Check if this key is in the database...
            final UserAPIAccount acc = DatabaseManager.getAPIAccount(key).block();
            if (acc != null && !acc.getBlocked()) {
                return new AuthenticationState(true)
                    .status(HttpStatusCodes.STATUS_CODE_OK)
                    .reason("Success")
                    .keyUsed(key)
                    .fromDisCalNetwork(false);
            }

            //If we reach here, the API key does not exist or is blocked...
            return new AuthenticationState(false)
                .status(GlobalConst.STATUS_AUTHORIZATION_DENIED)
                .reason("Authorization Denied");
        } else {
            LogFeed.log(LogObject
                .forDebug("Attempted API use without Authorization Header",
                    "IP: " + request.getRemoteAddr()));
            return new AuthenticationState(false)
                .status(GlobalConst.STATUS_BAD_REQUEST)
                .reason("Bad Request");
        }
    }

    public static void saveTempKey(final String key) {
        if (!tempKeys.containsKey(key))
            tempKeys.put(key, System.currentTimeMillis() + GlobalConst.oneDayMs);
    }

    public static void removeTempKey(final String key) {
        tempKeys.remove(key);
    }

    public static void saveReadOnlyKey(final String key) {
        if (!readOnlyKeys.containsKey(key)) {
            readOnlyKeys.put(key, System.currentTimeMillis() + GlobalConst.oneHourMs);
        }
    }

    //Timer handling
    public static void init() {
        timer = new Timer(true);
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                handleExpiredKeys();
            }
        }, GlobalConst.oneHourMs);
    }

    public static void shutdown() {
        if (timer != null)
            timer.cancel();
    }

    private static void handleExpiredKeys() {
        final List<String> allToRemove = new ArrayList<>();

        for (final String key : tempKeys.keySet()) {
            final long expireTime = tempKeys.get(key);
            if (expireTime >= System.currentTimeMillis())
                allToRemove.add(key);
        }

        for (final String key : readOnlyKeys.keySet()) {
            final long expireTime = readOnlyKeys.get(key);
            if (expireTime >= System.currentTimeMillis())
                allToRemove.add(key);
        }

        for (final String key : allToRemove) {
            tempKeys.remove(key);
            readOnlyKeys.remove(key);
        }
    }
}

File: server/src/main/kotlin/org/dreamexposure/discal/server/utils/Authentication.kt
package org.dreamexposure.discal.server.utils

import org.dreamexposure.discal.core.`object`.BotSettings
import org.dreamexposure.discal.core.`object`.web.AuthenticationState
import org.dreamexposure.discal.core.database.DatabaseManager
import org.dreamexposure.discal.core.logger.LOGGER
import org.dreamexposure.discal.core.utils.GlobalVal
import org.dreamexposure.discal.core.utils.GlobalVal.DEFAULT
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

//TODO: Use DI and make this an application runner as well
object Authentication {
    init {
        Flux.interval(Duration.ofHours(1))
            .map { handleExpiredKeys() }
            .subscribe()
    }

    //TODO: Switch to atomic containing immutable map eventually
    private val tempKeys: ConcurrentMap<String, Long> = ConcurrentHashMap()
    private val readOnlyKeys: ConcurrentMap<String, Long> = ConcurrentHashMap()

    fun authenticate(swe: ServerWebExchange): Mono<AuthenticationState> {
        //Check if correct method
        if (!swe.request.methodValue.equals("POST", true) || swe.request.methodValue.equals("GET", true)) {
            LOGGER.debug("Denied access | Method: ${swe.request.methodValue}")
            return Mono.just(AuthenticationState(false)
                    .status(GlobalVal.STATUS_NOT_ALLOWED)
                    .reason("Method not allowed")
            )
        }

        //Requires auth header
        if (swe.request.headers.getOrEmpty("Authorization").isNotEmpty()) {
            val authKey = swe.request.headers["Authorization"]!![0]

            return when {
                authKey == BotSettings.BOT_API_TOKEN.get() -> { //This is from within discal network
                    Mono.just(AuthenticationState(true)
                            .status(GlobalVal.STATUS_SUCCESS)
                            .reason("Success")
                            .keyUsed(authKey)
                            .fromDisCalNetwork(true)
                    )
                }
                tempKeys.containsKey(authKey) -> { //Temp key granted for logged in user
                    Mono.just(AuthenticationState(true)
                            .status(GlobalVal.STATUS_SUCCESS)
                            .reason("Success")
                            .keyUsed(authKey)
                            .fromDisCalNetwork(false)
                            .readOnly(false)
                    )
                }
                readOnlyKeys.containsKey(authKey) -> { //Read-only key granted for embed pages
                    Mono.just(AuthenticationState(true)
                            .status(GlobalVal.STATUS_SUCCESS)
                            .reason("Success")
                            .keyUsed(authKey)
                            .fromDisCalNetwork(false)
                            .readOnly(true)
                    )
                }
                authKey == "teapot" -> { //I'm a teapot
                    Mono.just(AuthenticationState(false)
                            .status(GlobalVal.STATUS_TEAPOT)
                            .reason("I'm a teapot")
                            .keyUsed(authKey)
                    )
                }
                else -> { //Check if key is in database...
                    DatabaseManager.getAPIAccount(authKey).map { acc ->
                        if (!acc.blocked) {
                            return@map AuthenticationState(true)
                                    .status(GlobalVal.STATUS_SUCCESS)
                                    .reason("Success")
                                    .keyUsed(authKey)
                                    .fromDisCalNetwork(false)
                        } else {
                            return@map AuthenticationState(false)
                                    .status(GlobalVal.STATUS_AUTHORIZATION_DENIED)
                                    .reason("Authorization Denied. API key blocked")
                        }
                    }.defaultIfEmpty(AuthenticationState(false)
                            .status(GlobalVal.STATUS_AUTHORIZATION_DENIED)
                            .reason("Authorization Denied")
                    )
                }
            }
        } else {
            LOGGER.debug(DEFAULT, "Attempted to use API without auth | ${swe.request.remoteAddress}")

            return Mono.just(AuthenticationState(false)
                    .status(GlobalVal.STATUS_BAD_REQUEST)
                    .reason("Bad Request")
            )
        }
    }

    fun saveTempKey(key: String) {
        if (!tempKeys.containsKey(key))
            tempKeys[key] = System.currentTimeMillis() + Duration.ofDays(1).toMillis()
    }

    fun removeTempKey(key: String) {
        tempKeys.remove(key)
    }

    fun saveReadOnlyKey(key: String) {
        if (!readOnlyKeys.containsKey(key))
            readOnlyKeys[key] = System.currentTimeMillis() + Duration.ofHours(1).toMillis()
    }

    private fun handleExpiredKeys() {
        val allToRemove = mutableListOf<String>()

        tempKeys.forEach { if (it.value >= System.currentTimeMillis()) allToRemove += it.key }
        readOnlyKeys.forEach { if (it.value >= System.currentTimeMillis()) allToRemove += it.key }
        allToRemove.forEach { tempKeys.remove(it);readOnlyKeys.remove(it) }
    }
}
